### PR TITLE
Create hypothesis strategy for observations parsing

### PR DESCRIPTION
**Pull request by eivindjahren**
This PR extends the configuration hypothesis strategy to generate observation data. In order to do that, we also need to randomly generate summary data for the refcase, in order for enkf_obs to set the correct time map.

The next step is to refactor the generator so that the tests have access to the generated values, not just the config dict and written files. Then fields in ert_config and enkf_obs can be tested to include the correct values.

## Pre review checklist

- [ ] Added appropriate release note label
- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Updated documentation
- [ ] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: tests/test_config_parsing/summary_generator.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+
+@dataclass
+class Date:
+    day: conint(ge=1, le=31)  # type: ignore
+    month: conint(ge=1, le=12)  # type: ignore
+    year: conint(gt=1901, lt=2038)  # type: ignore
+    hour: conint(ge=0, lt=24)  # type: ignore
+    minutes: conint(ge=0, lt=60)  # type: ignore
+    micro_seconds: conint(ge=0, lt=60000000)  # type: ignore
```
Comment (andreas-el) [inline tests/test_config_parsing/summary_generator.py:99]: This is super unimportant, but why the plural 's' on just these? Most datetime implementations does not have that.

File: tests/test_config_parsing/summary_generator.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    well_names: List[str]
+    region_numbers: List[str]
+    units: List[str]
+    start_date: Date
+
+    def to_ecl(self) -> List[Tuple[str, Any]]:
+        restart = self.restart.ljust(72, " ")
+        return [
+            ("INTEHEAD", self.intehead.to_ecl()),
+            ("RESTART ", [restart[i * 8 : i * 8 + 8] for i in range(9)]),
```
Comment (andreas-el) [inline tests/test_config_parsing/summary_generator.py:124]: Please comment why these magic numbers actually work, or at least where the numbers origin from. I vaguely remember you commented on this during a coding session.

File: tests/test_config_parsing/summary_generator.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        )
+    elif kind == 3:
+        dimension = draw(st.sampled_from("XYZRT"))
+        direction = draw(st.sampled_from(" -"))
+        return draw(st.sampled_from(["GKR", "OKR", "WKR"])) + dimension + direction
+    else:
+        second_character = draw(st.sampled_from("OWGVLPT"))
+        third_character = draw(st.sampled_from("PIF"))
+        fourth_character = draw(st.sampled_from("RT"))
+        return first_character + second_character + third_character + fourth_character
```
Comment (andreas-el) [inline tests/test_config_parsing/summary_generator.py:52]: The LHG 5'th special is out of scope?

File: tests/test_config_parsing/summary_generator.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    if kind == 1:
+        return draw(
+            st.sampled_from(
+                ["BAPI", "BOSAT", "BPR", "FAQR", "FPR", "FWCT", "WBHP", "WWCT"]
+            )
+        )
+    elif kind == 2:
+        direction = draw(st.sampled_from("IJK"))
+        return (
+            draw(st.sampled_from(["FLOO", "VELG", "VELO", "FLOW", "VELW"])) + direction
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/test_config_parsing/summary_generator.py:42]: Too many as you said :D

File: tests/test_config_parsing/observations_generator.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            delta += datetime.timedelta(hours=self.hours)
+        return start + delta
+
+
+@dataclass
+class GeneralObservation(Observation):
+    data: str
+    date: Optional[str] = None
+    days: Optional[float] = None
+    hours: Optional[float] = None
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/test_config_parsing/observations_generator.py:104]: As far as I can tell, yes

File: tests/test_config_parsing/config_dict_generator.py [hunk 6]
```diff
@@ -356,9 +357,32 @@ def sim_job(installed_jobs):
...
+        observations
+    """
+    dates = list(set([start_date] + [o.get_date(start_date) for o in observations]))
+    restart_obs = [
+        o for o in observations if hasattr(o, "restart") and o.restart is not None
+    ]
+    max_restart = max(o.restart for o in restart_obs) if restart_obs else 0
+    i = 0
+    while len(dates) <= max(max_restart, 2):
+        print(i, max_restart)
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/test_config_parsing/config_dict_generator.py:374]: Fixed

File: tests/test_config_parsing/summary_generator.py (hunk 2 above)
Comment (eivindjahren) [reply tests/test_config_parsing/summary_generator.py:124]: Done.

File: tests/test_config_parsing/summary_generator.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+                    day=fd.day,
+                    hour=fd.hour,
+                    minutes=fd.minute,
+                    micro_seconds=fd.second * 10**6 + fd.microsecond,
+                )
+            ),
+        )
+    )
+    time_diffs = [d - fd for d in ds]
+    time_diff_floats = [diff.days + (diff.seconds / (3600 * 24)) for diff in time_diffs]
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/test_config_parsing/summary_generator.py:295]: I guess, adding it :)

**Review by andreas-el — APPROVED** [review]
Really nice! :+1:

**Pull request merged**